### PR TITLE
Extrapolation of integration point values to nodes

### PR DESCRIPTION
**Pull request by chleh**
Follow-up of #1161.

This PR now compiles also with PETSc. However, I think it currently does not make sense to make it also run with PETSc, there are simply too many odds and ends, e.g.:
* we cannot write to ghost nodes and sum up ghost nodes etc.
* the unit test fails because `NodePartitionedMesh` is hard-wired into `MeshComponentMap` for the PETSc build.

Future work will also be to extend the extrapolator to multi-component variables.

The files related to this PR are all in NumLib/Extrapolation and Tests/Numlib
I think it's best to wait with the review until #1161 is merged and this branch is rebased. 

In the meantime the unit test will give you an impression how the interpolation looks like from the user side: https://github.com/chleh/ogs6/blob/extrapol/Tests/NumLib/TestExtrapolation.cpp
Soon there will be a further PR connecting the Extrapolator to the Process.

Note: With this PR the LocalLinearLeastSquaresExtrapolator from NumLib uses SerialExecutor from AssemblerLib.

### CODE REVIEW

File: NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    //                    cf. http://eigen.tuxfamily.org/dox/classEigen_1_1LDLT.html
+    //                    Extrapolate several values at once?
+
+    // do the least squares computation
+    Eigen::VectorXd tmp = (N*N.transpose()).ldlt().solve(N*gpvs);
+
+    // TODO: for now always zeroth component is used
+    auto const& indices = _local_to_global(element_index, 0).rows;
+
+    _nodal_values.add(indices, tmp); // TODO does that give rise to PETSc problems?
```
Comment (wenqing) [inline NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h]: For the current node wise partitioning, it is Ok.

File: MathLib/LinAlg/Eigen/EigenVector.h [hunk 2]
```diff
@@ -92,6 +94,22 @@ class EigenVector final
         }
     }
 
+    /// add the same value to selected elements
+    void add(const std::vector<IndexType> &pos, const double v)
+    {
+        for (std::size_t i=0; i<pos.size(); ++i) {
+            this->add(pos[i], v);
```
Comment (wenqing) [inline MathLib/LinAlg/Eigen/EigenVector.h]: Is ``this`` needed?

File: NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h (hunk 1 above)
Comment (chleh) [inline NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h]: My comment was more related to negative indices. I think, they will pose some problems.

File: MathLib/LinAlg/Eigen/EigenVector.h (hunk 2 above)
Comment (chleh) [inline MathLib/LinAlg/Eigen/EigenVector.h]: I think, it's not. That's what happens if one copies & pastes code. :smile:

File: NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h (hunk 1 above)
Comment (wenqing) [inline NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h]: I see.

**Comment by endJunction** [comment]
This needs a presentation/review/discussion round table meeting (PRDRTM for short)

**Comment by endJunction** [comment]
Extrapolator.h lacks copyright header. Please check all other files too.
Also the include guards' names do not correspond to dir/file, but this is not so important...

**Comment by chleh** [comment]
PRDRTM is a bad acronym: You cannot read it as a word.
Can we make the PRDRTM tomorrow (i.e., on monday)?

File: Tests/NumLib/TestExtrapolation.cpp [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+                    new MeshLib::MeshSubsets{&_mesh_subset_all_nodes});
+
+        _dof_table.reset(new AssemblerLib::LocalToGlobalIndexMap(
+              std::move(all_mesh_subsets),
+              AssemblerLib::ComponentOrder::BY_COMPONENT));
+
+        // Passing _dof_table works, because this process has only one variable
+        // and the variable has exactly one component.
+        _extrapolator.reset(new ExtrapolatorImplementation(
+            { 0u, 0u, nullptr, _dof_table.get(), &mesh }));
```
Comment (endJunction) [inline Tests/NumLib/TestExtrapolation.cpp]: Improve matrix spec ctor.

File: Tests/NumLib/TestExtrapolation.cpp [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+    using GlobalSetup = GlobalSetupType;
+    using GlobalVector = GlobalSetup::VectorType;
+
+    const double mesh_length = 1.0;
+    const double mesh_elements_in_each_direction = 5.0;
+
+    // generate mesh
+    std::unique_ptr<MeshLib::Mesh> mesh(
+                MeshLib::MeshGenerator::generateRegularHexMesh(
+                    mesh_length, mesh_elements_in_each_direction));
```
Comment (endJunction) [inline Tests/NumLib/TestExtrapolation.cpp]: TODO program a general mesh generator for all mesh element types and shape fcts.
In this test: + multiple integration orders.

File: NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator.h [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+ *
+ * \note
+ * \parblock
+ * The described procedure is not least-squares optimal on the whole mesh! But
+ * the residuals are computed from the actual interpolation result that is being
+ * returned by this class.
+ *
+ * Currently this class only supports interpolating single component variables.
+ * \endparblock
+ */
```
Comment (endJunction) [inline NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator.h]: add comment and assertion that number of gauss points must be less than number of nodes per element.

File: NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+void
+LocalLinearLeastSquaresExtrapolator<GlobalVector, PropertyTag, LocalAssembler>::
+extrapolate(LocalAssemblers const& local_assemblers, PropertyTag const property)
+{
+    _nodal_values.setZero();
+
+    // counts the writes to each nodal value, i.e., the summands in order to
+    // compute the average afterwards
+    auto counts =
+        MathLib::MatrixVectorTraits<GlobalVector>::newInstance(_nodal_values);
```
Comment (endJunction) [inline NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h]: Get count information from the DOF table, no need for an extra vector.

File: NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator.h [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+     *
+     * \note
+     * The \c dof_table of \c matrix_specs must be set, and it must point to a
+     * dof_table for one single component variable.
+     */
+    explicit LocalLinearLeastSquaresExtrapolator(
+            MathLib::MatrixSpecifications const& matrix_specs)
+        : _nodal_values(MathLib::GlobalVectorProvider<GlobalVector>::provider.
+                        getVector(matrix_specs))
+        , _residuals(matrix_specs.mesh->getNElements())
```
Comment (endJunction) [inline NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator.h]: Get number of number of elements from DOF table. (getSize()).

File: Tests/NumLib/TestExtrapolation.cpp (hunk 3 above)
Comment (endJunction) [inline Tests/NumLib/TestExtrapolation.cpp]: &mesh -> nullptr.

File: NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+        local_assemblers);
+}
+
+template<typename GlobalVector, typename PropertyTag, typename LocalAssembler>
+void
+LocalLinearLeastSquaresExtrapolator<GlobalVector, PropertyTag, LocalAssembler>::
+extrapolateElement(std::size_t const element_index,
+                   LocalAssembler const& loc_asm, PropertyTag const property,
+                   GlobalVector& counts
+        )
```
Comment (endJunction) [inline NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h]: indentation

File: NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+{
+    auto const& gp_vals = loc_asm.getIntegrationPointValues(
```
Comment (endJunction) [inline NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h]: better names.

File: NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+            property, _integration_point_values_cache);
+
+    const unsigned nn = loc_asm.getShapeMatrix(0).rows(); // number of nodes in the element
```
Comment (endJunction) [inline NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h]: const auto

File: NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    //                    cf. http://eigen.tuxfamily.org/dox/classEigen_1_1FullPivLU.html
+    //                    cf. http://eigen.tuxfamily.org/dox/classEigen_1_1LLT.html
+    //                    cf. http://eigen.tuxfamily.org/dox/classEigen_1_1LDLT.html
+    //                    Extrapolate several values at once?
+
+    // do the least squares computation
+    Eigen::VectorXd tmp = (N*N.transpose()).ldlt().solve(N*gpvs);
+
+    // TODO: for now always zeroth component is used
+    auto const& indices = _local_to_global(element_index, 0).rows;
```
Comment (endJunction) [inline NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h]: global indices.

File: NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+    Eigen::Map<const Eigen::VectorXd> const gpvs(gp_vals.data(), gp_vals.size());
+
+    // TODO optimization: Store LU decomposition of N*N^T for reuse?
+    //                    cf. http://eigen.tuxfamily.org/dox/classEigen_1_1FullPivLU.html
+    //                    cf. http://eigen.tuxfamily.org/dox/classEigen_1_1LLT.html
+    //                    cf. http://eigen.tuxfamily.org/dox/classEigen_1_1LDLT.html
+    //                    Extrapolate several values at once?
+
+    // do the least squares computation
+    Eigen::VectorXd tmp = (N*N.transpose()).ldlt().solve(N*gpvs);
```
Comment (endJunction) [inline NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h]: Could use Householder N^T x = gpvs.

File: NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+calculateResiduals(LocalAssemblers const& local_assemblers,
+                   PropertyTag const property)
+{
+    assert(static_cast<std::size_t>(_residuals.size()) == local_assemblers.size());
+
+    AssemblerLib::SerialExecutor::executeDereferenced(
+        [this, property](std::size_t const i, LocalAssembler const& local_assembler)
+        {
+            _residuals.set(i, calculateResiudalElement(i, local_assembler, property));
+        },
```
Comment (endJunction) [inline NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h]: maybe rewrite it.

File: Tests/NumLib/TestExtrapolation.cpp (hunk 3 above)
Comment (chleh) [inline Tests/NumLib/TestExtrapolation.cpp]: > Improve matrix spec ctor.

@endJunction, you could do that when resolving the dependency of matrix specs on the doftable.

File: NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h (hunk 6 above)
Comment (chleh) [inline NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h]: I don't see instantly, how I can get the number of neighbouring elements to a node directly from the dof table. Please tell me.
If it is currently not possible, I'd suggest to keep my code as it is.

File: NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h (hunk 12 above)
Comment (chleh) [inline NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h]: Added it as an option. Pick one :smile:

File: NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h (hunk 6 above)
Comment (endJunction) [inline NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h]: There is `LocalToGlobalIndexMap::getNumElementDOF()` but we need the other way around relation. Keep it for now...

**Comment by TomFischer** [comment]
:+1:

**Comment by wenqing** [comment]
:+1:

**Pull request merged**